api/odl.py

```diff
@@ -0,0 +1,598 @@
+        if loan:
+            url = loan.external_identifier
+        else:
+            checkout_id = str(uuid.uuid1())
+            expires = datetime.datetime.utcnow() + datetime.timedelta(days=self.default_loan_period)
+            # The patron UUID is generated randomly on each loan, so the distributor
+            # doesn't know when multiple loans come from the same patron.
+            patron_id = str(uuid.uuid1())
+            notification_url = self._url_for(
+                "odl_notify",
+                library_short_name=patron.library.short_name,
+                loan_external_identifier=checkout_id,
+                _external=True,
+            )
+
+            params = dict(
+                url=self.consolidated_loan_url,
+                id=id,
+                checkout_id=checkout_id,
+                patron_id=patron_id,
+                expires=(expires.isoformat() + 'Z'),
+                notification_url=notification_url,
+            )
+            url = "%(url)s?id=%(id)s&checkout_id=%(checkout_id)s&patron_id=%(patron_id)s&expires=%(expires)s&notification_url=%(notification_url)s" % params
+
+        response = self._get(url)
+
+        try:
+            status_doc = json.loads(response.content)
+        except ValueError, e:
+            raise BadResponseException(url, "License Status Document was not valid JSON.")
+        if status_doc.get("status") not in self.STATUS_VALUES:
+            raise BadResponseException(url, "License Status Document had an unknown status value.")
+        return status_doc
+
+    def checkin(self, patron, pin, licensepool):
+        """Return a loan early."""
+        _db = Session.object_session(patron)
+
+        loan = _db.query(Loan).filter(
+            Loan.patron==patron
+        ).filter(
+            Loan.license_pool_id==licensepool.id
+        )
+        if loan.count() < 1:
+            raise NotCheckedOut()
+        loan = loan.one()
+
+        doc = self.get_license_status_document(licensepool, patron, loan)
+        status = doc.get("status")
+        if status in [self.REVOKED_STATUS, self.RETURNED_STATUS, self.CANCELLED_STATUS, self.EXPIRED_STATUS]:
+            # This loan was already returned early or revoked by the distributor, or it expired.
+            raise NotCheckedOut()
+        elif status == self.ACTIVE_STATUS:
+            # This loan has already been fulfilled, so it needs to be returned through the DRM system.
+            raise CannotReturn()
+
+        return_url = doc.get("links", {}).get("return", {}).get("href")
+        if return_url:
+            # Hit the distributor's return link.
+            self._get(return_url)
+            # Get the status document again to make sure the return was successful,
+            # and if so update the pool availability and delete the local loan.
+            self.update_loan(loan)
+        else:
+            # The distributor didn't provide a link to return this loan.
+            raise CannotReturn()
```

Comment:
Whereas this I think is a good place to raise `CannotReturn` -- there's something wrong with the remote server.